components/script/dom/document.rs

```diff
@@ -2399,10 +2401,98 @@ impl Document {
         }
     }
 
+    /// <https://drafts.csswg.org/cssom-view/#document-run-the-scroll-steps>
+    pub(crate) fn run_the_scroll_steps(&self, can_gc: CanGc) {
+        // Step 1.
+        // > Run the steps to dispatch pending scrollsnapchanging events for doc.
+        // TODO(#7673): Implement scroll snapping
+
+        // Step 2
+        // > For each item target in doc’s pending scroll event targets, in the order they
+        // > were added to the list, run these substeps:
+        for target in self.pending_scroll_event_targets.borrow().iter() {
+            // Step 2.1
+            // > If target is a Document, fire an event named scroll that bubbles at target.
+            if target.downcast::<Document>().is_some() {
+                target.fire_bubbling_event(Atom::from("scroll"), can_gc);
+            }
+
+            // Step 2.2
+            // > Otherwise, fire an event named scroll at target.
+            if target.downcast::<Document>().is_some() {
+                target.fire_bubbling_event(Atom::from("scroll"), can_gc);
+            }
+        }
+
+        // Step 3.
+        // > Empty doc’s pending scroll event targets.
+        self.pending_scroll_event_targets.borrow_mut().clear();
+
+        // Step 4.
+        // > Run the steps to dispatch pending scrollsnapchange events for doc.
+        // TODO(#7673): Implement scroll snapping
+    }
+
+    /// Step 2-4 of viewport scroll event.
+    /// <https://drafts.csswg.org/cssom-view/#scrolling-events>
+    pub(crate) fn handle_viewport_scroll_event(&self) {
+        // Step 2.
+        // > If doc is a snap container, run the steps to update scrollsnapchanging targets
+        // > for doc with doc’s eventual snap target in the block axis as newBlockTarget and
+        // > doc’s eventual snap target in the inline axis as newInlineTarget.
+        // TODO(#7673): Implement scroll snapping
+
+        // Step 3.
+        // > If doc is already in doc’s pending scroll event targets, abort these steps.
+        let target = self.upcast::<EventTarget>();
+        if self
+            .pending_scroll_event_targets
+            .borrow()
+            .contains(&Dom::from_ref(target))
```

Comment:
I would prefer to use `any` with derefs to compare the underlying objects. I think this is technically safe but I don't like the idea of creating a `Dom` and not actually storing it anywhere.

Reply:
It does looks much more neat, thanks!